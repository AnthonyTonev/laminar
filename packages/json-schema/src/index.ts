export { ValidationError } from './ValidationError';
export { ResolveError } from './ResolveError';
export * as validation from './validation';
export { Schema } from './schema';
export { ResolvedSchema, resolve, resolveFile } from './resolve';
export {
  validate,
  validateCompiled,
  ensureValid,
  isCompiled,
  compile,
  ValidateOptions,
  ValidateResult,
} from './validate';
