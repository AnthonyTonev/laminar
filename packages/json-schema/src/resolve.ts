import { existsSync, readFileSync } from 'fs';
import * as YAML from 'yaml';
import fetch from 'node-fetch';
import { dirname, join } from 'path';
import { URL } from 'url';
import { ResolveError } from './ResolveError';
import { Schema } from './schema';

export interface RefMap {
  [ref: string]: Schema;
}

export interface ResolvedSchema<T extends Schema = Schema> {
  schema: T;
  refs: RefMap;
  uris: string[];
}

interface LoadedJsonObject {
  uri: string;
  content: Schema;
  cwd?: string;
}

interface TraversableSchema {
  [key: string]: Schema;
}

interface RefSchema {
  $ref: string;
  [key: string]: unknown;
}

interface RecursiveRefSchema {
  $recursiveRef: string;
  [key: string]: unknown;
}

interface Context {
  schema: Schema;
  refs: RefMap;
}

interface FileContext {
  parentId?: string;
  cwd?: string;
  uris?: string[];
}

const isTraversable = (schema: unknown): schema is TraversableSchema => Boolean(schema && typeof schema === 'object');

const toUrl = (url: string, base?: string): string | undefined => {
  try {
    return new URL(url, base).toString();
  } catch (error) {
    return undefined;
  }
};

const getId = (schema: Schema): string | undefined => {
  if (isTraversable(schema)) {
    if ('$id' in schema && schema.$id && typeof schema.$id === 'string') {
      return schema.$id;
    } else if ('id' in schema && schema.id && typeof schema.id === 'string') {
      return schema.id;
    }
  }
  return undefined;
};

const getAnchor = (schema: Schema): string | undefined =>
  isTraversable(schema) && '$anchor' in schema && schema.$anchor && typeof schema.$anchor === 'string'
    ? schema.$anchor
    : undefined;

const isRefSchema = (schema: Schema): schema is RefSchema =>
  isTraversable(schema) && '$ref' in schema && schema.$ref !== undefined && typeof schema.$ref === 'string';

const isRecursiveRefSchema = (schema: Schema): schema is RecursiveRefSchema =>
  isTraversable(schema) &&
  '$recursiveRef' in schema &&
  schema.$recursiveRef !== undefined &&
  typeof schema.$recursiveRef === 'string';

const currentId = (schema: Schema, parentId?: string): string | undefined => {
  const id = getId(schema);
  return id ? toUrl(id, parentId) : parentId;
};

const currentUrl = (url?: string, { cwd, parentId }: FileContext = {}): string | undefined => {
  const fullUrl = url && toUrl(url, parentId);
  if (fullUrl) {
    return fullUrl;
  } else if (url && cwd) {
    if (!existsSync(join(cwd, url))) {
      throw new ResolveError(`File ${url} in ${cwd} does not exist`);
    }
    return url;
  } else {
    return undefined;
  }
};

const reduceSchema = <TResult = Schema>(
  schema: Schema,
  cb: (all: TResult, item: Schema, id?: string) => TResult,
  initial: TResult,
  id?: string,
): TResult =>
  isTraversable(schema)
    ? Object.values(schema).reduce(
        (all, item) => reduceSchema(item, cb, all, currentId(schema, id)),
        cb(initial, schema, id),
      )
    : cb(initial, schema, id);

export const extractNamedRefs = (document: Schema): RefMap =>
  reduceSchema(
    document,
    (all, item, id) => {
      const itemId = getId(item);
      const anchorId = getAnchor(item);
      const url = anchorId
        ? [itemId ? toUrl(itemId, id) : id, anchorId].join('#')
        : itemId
        ? toUrl(itemId, id) ?? itemId
        : undefined;
      return url ? { ...all, [url]: item } : all;
    },
    {},
  );

export const extractUrls = (schema: Schema, namedRefs: string[] = [], fileContext: FileContext = {}): string[] => {
  const namedRefsAsUrls = namedRefs.map((ref) => toUrl(ref)?.split('#')[0] ?? ref);
  return reduceSchema<string[]>(
    schema,
    (all, item, parentId) => {
      if (isRefSchema(item) || isRecursiveRefSchema(item)) {
        const [url] = (isRefSchema(item) ? item.$ref : item.$recursiveRef).split('#');
        const fullUrl = currentUrl(url, { ...fileContext, parentId });
        if (fullUrl && !all.includes(fullUrl) && !namedRefsAsUrls.includes(fullUrl)) {
          return [...all, fullUrl];
        }
      }
      return all;
    },
    [],
  );
};

const loadFile = async (uri: string, { cwd }: FileContext = {}): Promise<LoadedJsonObject> => {
  const url = toUrl(uri);
  if (url) {
    const result = await fetch(uri);
    if (result.headers.get('content-type') === 'application/yaml') {
      return { uri, content: YAML.parse(await result.text()) };
    } else {
      return { uri, content: await result.json() };
    }
  } else {
    const file = cwd ? join(cwd, uri) : uri;
    const content = readFileSync(file, 'utf8');
    const newCwd = dirname(file);

    if (uri.endsWith('.yaml') || uri.endsWith('.yml')) {
      return { uri: `file://${file}`, content: YAML.parse(content), cwd: newCwd };
    } else {
      return { uri: `file://${file}`, content: JSON.parse(content), cwd: newCwd };
    }
  }
};

const parseJsonPointer = (name: string): string => decodeURIComponent(name.replace('~1', '/').replace('~0', '~'));

const getJsonPointer = (document: Schema, pointer: string): Schema | undefined =>
  pointer
    .split('/')
    .reduce<Schema | undefined>(
      (item, name) => (name ? (isTraversable(item) ? item[parseJsonPointer(name)] : undefined) : item),
      document,
    );

const resolveNestedRefs = (schema: Schema, context: Context, fileContext: FileContext = {}): Schema => {
  const parentId = currentId(schema, fileContext.parentId);

  if (isTraversable(schema)) {
    for (const [key, item] of Object.entries(schema)) {
      schema[key] = resolveNestedRefs(item, context, { ...fileContext, parentId });
    }
  }

  if (isRecursiveRefSchema(schema)) {
    const { $recursiveRef, ...restSiblings } = schema as RecursiveRefSchema;
    const [url, pointer] = $recursiveRef.split('#');
    const fullUrl = currentUrl(url, { ...fileContext, parentId });

    const currentDocument = fullUrl ? context.refs[fullUrl] : context.schema;
    const newContent = pointer ? getJsonPointer(currentDocument, pointer) : currentDocument;
    if (
      newContent !== undefined &&
      typeof newContent !== 'boolean' &&
      newContent.$id &&
      newContent.$recursiveAnchor === true
    ) {
      const anchorPointer = [newContent.$id, pointer].join('#');
      if (!context.refs[anchorPointer]) {
        context.refs[anchorPointer] = newContent;
      }
      return { $ref: anchorPointer, ...restSiblings };
    }
  } else if (isRefSchema(schema)) {
    const { $ref, ...restSiblings } = schema as RefSchema;
    const [url, pointer] = $ref.split('#');

    const fullUrl = currentUrl(url, { ...fileContext, parentId });
    const fullPointer = [fullUrl, pointer].join('#');

    if (!context.refs[fullPointer]) {
      const currentDocument = fullUrl ? context.refs[fullUrl] : context.schema;
      const newContent = pointer ? getJsonPointer(currentDocument, pointer) : currentDocument;
      if (newContent !== undefined) {
        context.refs[fullPointer] = newContent;
      }
    }

    return { $ref: fullPointer, ...restSiblings };
  }

  return schema;
};

export const extractFiles = async (
  schema: Schema,
  options: FileContext = {},
): Promise<{ refs: RefMap; uris: string[] }> => {
  const initialRefs = extractNamedRefs(schema);
  const result = await Promise.all(
    extractUrls(schema, Object.keys(initialRefs), options).map(async (url) => {
      const { content, cwd, uri } = await loadFile(url, options);
      const { refs, uris } = await extractFiles(content, { cwd });
      const ref = resolveNestedRefs(content, { schema: content, refs }, { cwd });
      return { uris: [...uris, uri], refs: { [url]: ref, ...refs } };
    }),
  );

  return result.reduce((all, item) => ({ uris: [...all.uris, ...item.uris], refs: { ...all.refs, ...item.refs } }), {
    uris: [],
    refs: initialRefs,
  });
};

export const resolve = async (original: Schema, fileContext: FileContext = {}): Promise<ResolvedSchema> => {
  const copy: Schema = JSON.parse(JSON.stringify(original));
  const { refs, uris } = await extractFiles(copy, fileContext);
  const context = { schema: copy, refs, uris };
  const schema = resolveNestedRefs(copy, context, fileContext);
  return { schema, refs, uris };
};

export const resolveFile = async (file: string): Promise<ResolvedSchema> => {
  const { content, cwd, uri } = await loadFile(file);
  const resolved = await resolve(content, { cwd });
  return { ...resolved, uris: [uri, ...resolved.uris] };
};
