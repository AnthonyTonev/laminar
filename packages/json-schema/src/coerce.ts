import { coercers, coerceSchema } from './coercion';
import { compile, isCompiled, ResolvedSchema } from './resolve';
import { Schema } from './schema';
import { drafts, Drafts } from './validate';

interface CoerceOptionsBase {
  value: unknown;
  name?: string;
  draft?: keyof Drafts;
  type?: keyof typeof coercers;
}

export interface CoerceSchemaOptions extends CoerceOptionsBase {
  schema: Schema | string;
}

export interface CoerceCompiledOptions extends CoerceOptionsBase {
  schema: ResolvedSchema;
}

export type CoerceOptions = CoerceSchemaOptions | CoerceCompiledOptions;

export function coerceCompiled({
  schema,
  value,
  name = 'value',
  draft = 'draft2019-09',
  type = 'json',
}: CoerceCompiledOptions): unknown {
  const validators = drafts[draft];
  return coerceSchema(schema.schema, value, { validators, refs: schema.refs, name, coercers: coercers[type] });
}

/**
 * Coerce a given value, using the provided json schema.
 *
 *  - With type: 'json'
 *    This is used to convert json validated with json schema into javascript objects.
 *    Namely it converts all the strings with format date and date-time into Date objects
 *  - With type: 'query'
 *    To convert a value coming from a URL query string to the type you want it to be,
 *    for example '12' with type: 'integer' will be converted to 12 so the validation can succeed.
 *
 * Additionally, we assign default values where appropriate.
 */
export function coerce(options: CoerceCompiledOptions): unknown;
export function coerce(options: CoerceSchemaOptions): Promise<unknown>;
export function coerce({ schema, ...rest }: CoerceOptions): unknown {
  return isCompiled(schema)
    ? coerceCompiled({ ...rest, schema })
    : compile(schema).then((compiledSchema) => coerceCompiled({ ...rest, schema: compiledSchema }));
}
