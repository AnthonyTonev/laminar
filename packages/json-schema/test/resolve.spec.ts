import nock from 'nock';
import { Schema } from '../src';
import { extractUrls, RefMap, extractNamedRefs, extractFiles, resolve } from '../src/resolve';

describe('json-refs', () => {
  it.each<[string, Schema, string[]]>([
    ['no ref', { $ref: '#/$defs/test', $defs: { test: { maximum: 10 } } }, []],
    ['single ref', { $ref: 'http://one.test#/test' }, ['http://one.test/']],
    [
      'nested multiple ref',
      {
        properties: {
          test: { $ref: 'http://one.test#/test' },
          other: { properties: { deep: { $ref: 'http://two.test#/test' } } },
        },
      },
      ['http://one.test/', 'http://two.test/'],
    ],
    [
      'nested multiple ref with ids',
      {
        $id: 'http://one.test',
        properties: {
          test: { $ref: 'first#/test' },
          other: { properties: { deep: { $ref: 'second#/test' } } },
        },
      },
      ['http://one.test/first', 'http://one.test/second'],
    ],
    [
      'nested multiple ref with ids old style',
      {
        id: 'http://one.test',
        properties: {
          test: { $ref: 'first#/test' },
          other: { properties: { deep: { $ref: 'second#/test' } } },
        },
      },
      ['http://one.test/first', 'http://one.test/second'],
    ],
    [
      'nested multiple refs and different ids',
      {
        properties: {
          one: {
            $id: 'http://one.test',
            properties: { test: { $ref: 'first#/test-1' } },
          },
          two: {
            $id: 'http://two.test',
            properties: { deep: { properties: { other: { $ref: 'second#/test-2' } } } },
          },
          standAlone: { $ref: 'http://three.test' },
        },
      },
      ['http://one.test/first', 'http://two.test/second', 'http://three.test/'],
    ],
  ])('Should extractUrls for %s', (_, schema, expected) => {
    expect(extractUrls(schema)).toEqual(expected);
  });

  it.each<[string, Schema, RefMap]>([
    [
      'single id',
      { $id: 'http://one.test', minimum: 12 },
      { 'http://one.test/': { $id: 'http://one.test', minimum: 12 } },
    ],
    [
      'nested id',
      {
        $id: 'http://one.test',
        minimum: 12,
        properties: { deep: { $id: 'other', maximum: 22 } },
      },
      {
        'http://one.test/': {
          $id: 'http://one.test',
          minimum: 12,
          properties: { deep: { $id: 'other', maximum: 22 } },
        },
        'http://one.test/other': { $id: 'other', maximum: 22 },
      },
    ],
  ])('Should extractNamedRefs for %s', (_, schema, expected) => {
    expect(extractNamedRefs(schema)).toEqual(expected);
  });

  it('Should extract multiple urls', async () => {
    const schema = {
      properties: {
        test: { $ref: 'http://one.test#/test' },
        other: {
          properties: {
            deep: { $ref: 'http://two.test/folder#/test' },
          },
        },
        last: {
          $id: 'http://two.test',
          properties: {
            deep: {
              $ref: 'folder',
            },
          },
        },
      },
    };

    nock('http://one.test').get('/').reply(200, { $ref: 'http://three.test#' });

    nock('http://two.test').get('/folder').reply(200, { test: 2 });

    nock('http://three.test').get('/').reply(200, { $ref: 'http://four.test#/test', test: 3 });

    nock('http://four.test').get('/').reply(200, { test: 4 });

    const expected = {
      refs: {
        'http://two.test/': { $id: 'http://two.test', properties: { deep: { $ref: 'folder' } } },
        'http://four.test/': { test: 4 },
        'http://four.test/#/test': 4,
        'http://three.test/': { $ref: 'http://four.test/#/test', test: 3 },
        'http://three.test/#': { $ref: 'http://four.test/#/test', test: 3 },
        'http://one.test/': { $ref: 'http://three.test/#' },
        'http://two.test/folder': { test: 2 },
      },
      uris: ['http://four.test/', 'http://three.test/', 'http://one.test/', 'http://two.test/folder'],
    };

    expect(await extractFiles(schema)).toEqual(expected);
  });

  it('Should resolve simple nested refs', async () => {
    const schema: Schema = {
      $defs: {
        a: { type: 'integer' as const },
        b: { $ref: '#/$defs/a' },
        c: { $ref: '#/$defs/b' },
      },
      $ref: '#/$defs/c',
    };

    const expected = {
      schema: {
        $ref: '#/$defs/c',
        $defs: {
          a: { type: 'integer' as const },
          b: { $ref: '#/$defs/a' },
          c: { $ref: '#/$defs/b' },
        },
      },
      refs: {
        '#/$defs/a': { type: 'integer' },
        '#/$defs/b': { $ref: '#/$defs/a' },
        '#/$defs/c': { $ref: '#/$defs/b' },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });

  it('Should resolve anchors with id', async () => {
    const schema = {
      $defs: {
        A: { $anchor: 'foo', $id: 'http://localhost:1234/bar', type: 'integer' as const },
      },
      $ref: 'http://localhost:1234/bar#foo',
    };

    const expected = {
      schema: {
        $ref: 'http://localhost:1234/bar#foo',
        $defs: {
          A: { $anchor: 'foo', $id: 'http://localhost:1234/bar', type: 'integer' as const },
        },
      },
      refs: {
        'http://localhost:1234/bar#foo': {
          $anchor: 'foo',
          $id: 'http://localhost:1234/bar',
          type: 'integer',
        },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });
  it('Should resolve anchors with nested id', async () => {
    const schema = {
      $defs: {
        A: { $defs: { B: { $anchor: 'foo', type: 'integer' as const } }, $id: 'nested.json' },
      },
      $id: 'http://localhost:1234/root',
      $ref: 'http://localhost:1234/nested.json#foo',
    };

    const expected = {
      schema: {
        $defs: {
          A: { $defs: { B: { $anchor: 'foo', type: 'integer' as const } }, $id: 'nested.json' },
        },
        $id: 'http://localhost:1234/root',
        $ref: 'http://localhost:1234/nested.json#foo',
      },
      refs: {
        'http://localhost:1234/root': {
          $defs: {
            A: { $defs: { B: { $anchor: 'foo', type: 'integer' as const } }, $id: 'nested.json' },
          },
          $id: 'http://localhost:1234/root',
          $ref: 'http://localhost:1234/nested.json#foo',
        },
        'http://localhost:1234/nested.json': {
          $defs: { B: { $anchor: 'foo', type: 'integer' } },
          $id: 'nested.json',
        },
        'http://localhost:1234/nested.json#foo': { $anchor: 'foo', type: 'integer' as const },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });

  it('Should resolve simple anchors', async () => {
    const schema = { $defs: { A: { $anchor: 'foo', type: 'integer' as const } }, $ref: '#foo' };

    const expected = {
      schema: {
        $defs: { A: { $anchor: 'foo', type: 'integer' as const } },
        $ref: '#foo',
      },
      refs: {
        '#foo': { $anchor: 'foo', type: 'integer' },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });

  it('Should resolve recursiveRef', async () => {
    const schema = {
      $id: 'http://localhost:1234',
      $recursiveAnchor: true,
      type: 'object' as const,
      properties: {
        myprop: {
          anyOf: [
            { type: 'string' as const },
            { type: 'object' as const, additionalProperties: { $recursiveRef: '#' } },
          ],
        },
      },
    };

    const expected = {
      schema: {
        $id: 'http://localhost:1234',
        $recursiveAnchor: true,
        type: 'object',
        properties: {
          myprop: {
            anyOf: [{ type: 'string' }, { type: 'object', additionalProperties: { $ref: 'http://localhost:1234#' } }],
          },
        },
      },
      refs: {
        'http://localhost:1234#': {
          $id: 'http://localhost:1234',
          $recursiveAnchor: true,
          type: 'object',
          properties: {
            myprop: {
              anyOf: [{ type: 'string' }, { type: 'object', additionalProperties: { $ref: 'http://localhost:1234#' } }],
            },
          },
        },
        'http://localhost:1234/': {
          $id: 'http://localhost:1234',
          $recursiveAnchor: true,
          type: 'object',
          properties: {
            myprop: {
              anyOf: [{ type: 'string' }, { type: 'object', additionalProperties: { $ref: 'http://localhost:1234#' } }],
            },
          },
        },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });

  it('Should resolve location independent definitiions', async () => {
    const schema = {
      allOf: [{ $ref: 'http://localhost:1234/bar#foo' }],
      $defs: { A: { $id: 'http://localhost:1234/bar#foo', type: 'integer' as const } },
    };

    const expected = {
      schema: {
        allOf: [{ $ref: 'http://localhost:1234/bar#foo' }],
        $defs: { A: { $id: 'http://localhost:1234/bar#foo', type: 'integer' } },
      },
      refs: {
        'http://localhost:1234/bar#foo': { $id: 'http://localhost:1234/bar#foo', type: 'integer' },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });

  it('Should resolve location independent definitiions with only url id', async () => {
    const schema: Schema = {
      allOf: [{ $ref: '#foo' }],
      $defs: { A: { $id: '#foo', type: 'integer' as const } },
    };

    const expected = {
      schema: {
        allOf: [{ $ref: '#foo' }],
        $defs: { A: { $id: '#foo', type: 'integer' } },
      },
      refs: {
        '#foo': { $id: '#foo', type: 'integer' },
      },
      uris: [],
    };

    expect(await resolve(schema)).toEqual(expected);
  });

  it('Should resolve nested recursive with files', async () => {
    const schema: Schema = {
      $id: 'http://localhost:1234/tree',
      description: 'tree of nodes',
      type: 'object',
      properties: {
        meta: { type: 'string' },
        nodes: {
          type: 'array',
          items: { $ref: 'node' },
        },
      },
      required: ['meta', 'nodes'],
      $defs: {
        node: {
          $id: 'http://localhost:1234/node',
          description: 'node',
          type: 'object' as const,
          properties: {
            value: { type: 'number' as const },
            subtree: { $ref: 'tree' },
          },
          required: ['value'],
        },
      },
    };

    expect(await resolve(schema)).toMatchSnapshot();
  });

  it('Should resolve complex refs', async () => {
    const schema: Schema = {
      type: 'object',
      properties: {
        mangled: { $ref: '#/$defs/unknown/more-unknown' },
        user: {
          type: 'object',
          properties: {
            name: { $ref: '#/$defs/title' },
            id: { $ref: '#/$defs/id' },
            address: { $ref: '#/$defs/address' },
          },
        },
      },
      $defs: {
        title: {
          type: 'string',
          pattern: '^[A-Z]',
        },
        address: {
          type: 'object',
          properties: {
            id: { $ref: '#/$defs/id' },
            street: { $ref: '#/$defs/title' },
            price: { $ref: '#/$defs/test~1slash' },
            town: { $ref: '#/$defs/test~0tilde' },
            num: { $ref: '#/$defs/test%25percent' },
          },
        },
        'test/slash': {
          type: 'number',
        },
        'test~tilde': {
          type: 'string',
        },
        'test%percent': {
          type: 'integer',
        },
        id: {
          type: 'string',
          minLength: 3,
          maxLength: 4,
        },
      },
    };

    expect(await resolve(schema)).toMatchSnapshot();
  });

  it('Should resolve url refs', async () => {
    nock('http://example.test')
      .get('/schema')
      .reply(200, { test: { type: 'number' } });

    const schema = {
      test: '123',
      properties: { other: { $ref: 'http://example.test/schema#/test' } },
    };
    expect(await resolve(schema)).toEqual({
      schema: { properties: { other: { $ref: 'http://example.test/schema#/test' } }, test: '123' },
      refs: {
        'http://example.test/schema': {
          test: {
            type: 'number',
          },
        },
        'http://example.test/schema#/test': {
          type: 'number',
        },
      },
      uris: ['http://example.test/schema'],
    });
  });

  it('Should resolve local refs from json', async () => {
    const schema = { properties: { test: { $ref: 'assets/test.json#/UserResponse' } } };

    expect(await resolve(schema, { cwd: __dirname })).toEqual({
      schema: { properties: { test: { $ref: 'assets/test.json#/UserResponse' } } },
      refs: {
        'assets/test.json': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'assets/test.json#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
      },
      uris: [expect.stringContaining('file://')],
    });
  });

  it('Should resolve local refs with file:://', async () => {
    const schema = { properties: { test: { $ref: 'file://./assets/test.json#/UserResponse' } } };

    expect(await resolve(schema, { cwd: __dirname })).toEqual({
      schema: { properties: { test: { $ref: 'file://./assets/test.json#/UserResponse' } } },
      refs: {
        'file://./assets/test.json': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'file://./assets/test.json#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
      },
      uris: [expect.stringContaining('file://')],
    });
  });

  it('Should resolve local refs from yaml', async () => {
    const schema = { properties: { test: { $ref: 'assets/test.yaml#/UserResponse' } } };

    expect(await resolve(schema, { cwd: __dirname })).toEqual({
      schema: { properties: { test: { $ref: 'assets/test.yaml#/UserResponse' } } },
      refs: {
        'assets/test.yaml': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'assets/test.yaml#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
      },
      uris: [expect.stringContaining('file://')],
    });
  });

  it('Should resolve local refs in nested files', async () => {
    const schema = {
      properties: { test: { $ref: 'assets/test.yml#/UserResponse' } },
    };

    expect(await resolve(schema, { cwd: __dirname })).toEqual({
      schema: { properties: { test: { $ref: 'assets/test.yml#/UserResponse' } } },
      refs: {
        'test.json': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'test.json#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
        'assets/test.yml': {
          UserResponse: {
            type: 'object',
            properties: { id: { $ref: 'test.json#/UserResponse' } },
          },
        },
        'assets/test.yml#/UserResponse': {
          type: 'object',
          properties: { id: { $ref: 'test.json#/UserResponse' } },
        },
      },
      uris: [expect.stringContaining('file://'), expect.stringContaining('file://')],
    });
  });

  it('Should resolve multiple refs with one load', async () => {
    nock('http://example.test')
      .get('/schema')
      .reply(200, { test: { type: 'number' }, other: { type: 'array' } });

    const schema = {
      properties: {
        test: { $ref: 'http://example.test/schema#/test' },
        other: { $ref: 'http://example.test/schema#/other' },
      },
    };

    expect(await resolve(schema)).toEqual({
      schema: {
        properties: {
          test: { $ref: 'http://example.test/schema#/test' },
          other: { $ref: 'http://example.test/schema#/other' },
        },
      },
      refs: {
        'http://example.test/schema': { test: { type: 'number' }, other: { type: 'array' } },
        'http://example.test/schema#/test': { type: 'number' },
        'http://example.test/schema#/other': { type: 'array' },
      },
      uris: ['http://example.test/schema'],
    });
  });

  it('Should resolve nested refs', async () => {
    nock('http://example.test')
      .get('/schema')
      .reply(200, { test: { type: 'number' }, other: { type: 'array' } })
      .get('/other-schema')
      .reply(200, { last: { type: 'object' } });

    const schema: Schema = {
      properties: {
        test: { $ref: 'http://example.test/schema#/test' },
        other: { $ref: 'http://example.test/schema#/other' },
        last: { $ref: 'http://example.test/other-schema#/last' },
      },
    };

    expect(await resolve(schema)).toEqual({
      schema: {
        properties: {
          test: { $ref: 'http://example.test/schema#/test' },
          other: { $ref: 'http://example.test/schema#/other' },
          last: { $ref: 'http://example.test/other-schema#/last' },
        },
      },
      refs: {
        'http://example.test/schema': { test: { type: 'number' }, other: { type: 'array' } },
        'http://example.test/other-schema': { last: { type: 'object' } },
        'http://example.test/schema#/test': { type: 'number' },
        'http://example.test/schema#/other': { type: 'array' },
        'http://example.test/other-schema#/last': { type: 'object' },
      },
      uris: ['http://example.test/schema', 'http://example.test/other-schema'],
    });
  });

  it('Should resolve recursive copy', async () => {
    const schema = { properties: { foo: { $ref: '#' } }, additionalProperties: false };

    const schema1 = await resolve(schema);
    const schema2 = await resolve(schema);

    expect(schema1).toEqual(schema2);
  });
});
