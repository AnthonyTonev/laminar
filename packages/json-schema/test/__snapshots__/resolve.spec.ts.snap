// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-refs Should resolve complex refs 1`] = `
Object {
  "refs": Object {
    "#/$defs/address": Object {
      "properties": Object {
        "id": Object {
          "$ref": "#/$defs/id",
        },
        "num": Object {
          "$ref": "#/$defs/test%25percent",
        },
        "price": Object {
          "$ref": "#/$defs/test~1slash",
        },
        "street": Object {
          "$ref": "#/$defs/title",
        },
        "town": Object {
          "$ref": "#/$defs/test~0tilde",
        },
      },
      "type": "object",
    },
    "#/$defs/id": Object {
      "maxLength": 4,
      "minLength": 3,
      "type": "string",
    },
    "#/$defs/test%25percent": Object {
      "type": "integer",
    },
    "#/$defs/test~0tilde": Object {
      "type": "string",
    },
    "#/$defs/test~1slash": Object {
      "type": "number",
    },
    "#/$defs/title": Object {
      "pattern": "^[A-Z]",
      "type": "string",
    },
  },
  "schema": Object {
    "$defs": Object {
      "address": Object {
        "properties": Object {
          "id": Object {
            "$ref": "#/$defs/id",
          },
          "num": Object {
            "$ref": "#/$defs/test%25percent",
          },
          "price": Object {
            "$ref": "#/$defs/test~1slash",
          },
          "street": Object {
            "$ref": "#/$defs/title",
          },
          "town": Object {
            "$ref": "#/$defs/test~0tilde",
          },
        },
        "type": "object",
      },
      "id": Object {
        "maxLength": 4,
        "minLength": 3,
        "type": "string",
      },
      "test%percent": Object {
        "type": "integer",
      },
      "test/slash": Object {
        "type": "number",
      },
      "test~tilde": Object {
        "type": "string",
      },
      "title": Object {
        "pattern": "^[A-Z]",
        "type": "string",
      },
    },
    "properties": Object {
      "mangled": Object {
        "$ref": "#/$defs/unknown/more-unknown",
      },
      "user": Object {
        "properties": Object {
          "address": Object {
            "$ref": "#/$defs/address",
          },
          "id": Object {
            "$ref": "#/$defs/id",
          },
          "name": Object {
            "$ref": "#/$defs/title",
          },
        },
        "type": "object",
      },
    },
    "type": "object",
  },
  "uris": Array [],
}
`;

exports[`json-refs Should resolve nested recursive with files 1`] = `
Object {
  "refs": Object {
    "http://localhost:1234/node": Object {
      "$id": "http://localhost:1234/node",
      "description": "node",
      "properties": Object {
        "subtree": Object {
          "$ref": "http://localhost:1234/tree#",
        },
        "value": Object {
          "type": "number",
        },
      },
      "required": Array [
        "value",
      ],
      "type": "object",
    },
    "http://localhost:1234/node#": Object {
      "$id": "http://localhost:1234/node",
      "description": "node",
      "properties": Object {
        "subtree": Object {
          "$ref": "http://localhost:1234/tree#",
        },
        "value": Object {
          "type": "number",
        },
      },
      "required": Array [
        "value",
      ],
      "type": "object",
    },
    "http://localhost:1234/tree": Object {
      "$defs": Object {
        "node": Object {
          "$id": "http://localhost:1234/node",
          "description": "node",
          "properties": Object {
            "subtree": Object {
              "$ref": "http://localhost:1234/tree#",
            },
            "value": Object {
              "type": "number",
            },
          },
          "required": Array [
            "value",
          ],
          "type": "object",
        },
      },
      "$id": "http://localhost:1234/tree",
      "description": "tree of nodes",
      "properties": Object {
        "meta": Object {
          "type": "string",
        },
        "nodes": Object {
          "items": Object {
            "$ref": "http://localhost:1234/node#",
          },
          "type": "array",
        },
      },
      "required": Array [
        "meta",
        "nodes",
      ],
      "type": "object",
    },
    "http://localhost:1234/tree#": Object {
      "$defs": Object {
        "node": Object {
          "$id": "http://localhost:1234/node",
          "description": "node",
          "properties": Object {
            "subtree": Object {
              "$ref": "http://localhost:1234/tree#",
            },
            "value": Object {
              "type": "number",
            },
          },
          "required": Array [
            "value",
          ],
          "type": "object",
        },
      },
      "$id": "http://localhost:1234/tree",
      "description": "tree of nodes",
      "properties": Object {
        "meta": Object {
          "type": "string",
        },
        "nodes": Object {
          "items": Object {
            "$ref": "http://localhost:1234/node#",
          },
          "type": "array",
        },
      },
      "required": Array [
        "meta",
        "nodes",
      ],
      "type": "object",
    },
  },
  "schema": Object {
    "$defs": Object {
      "node": Object {
        "$id": "http://localhost:1234/node",
        "description": "node",
        "properties": Object {
          "subtree": Object {
            "$ref": "http://localhost:1234/tree#",
          },
          "value": Object {
            "type": "number",
          },
        },
        "required": Array [
          "value",
        ],
        "type": "object",
      },
    },
    "$id": "http://localhost:1234/tree",
    "description": "tree of nodes",
    "properties": Object {
      "meta": Object {
        "type": "string",
      },
      "nodes": Object {
        "items": Object {
          "$ref": "http://localhost:1234/node#",
        },
        "type": "array",
      },
    },
    "required": Array [
      "meta",
      "nodes",
    ],
    "type": "object",
  },
  "uris": Array [],
}
`;
